### PR TITLE
Disable DatePicker vibrate on date change on iOS

### PR DESCRIPTION
**Pull request by cbracken**
In the Android native material DatePicker, haptic feedback is expected
on date change. This is not expected behaviour for iOS date/time
pickers.

### CODE REVIEW

**Comment by cbracken** [comment]
Fixes #8268.

**Comment by xster** [comment]
ship it

File: packages/flutter/lib/src/material/date_picker.dart [hunk 1]
```diff
@@ -657,6 +662,15 @@ class _DatePickerDialogState extends State<_DatePickerDialog> {
...
   Widget build(BuildContext context) {
+    switch (Theme.of(context).platform) {
+      case TargetPlatform.android:
+      case TargetPlatform.fuchsia:
+        _vibrate = true;
+        break;
+      case TargetPlatform.iOS:
+        _vibrate = false;
+        break;
+    }
```
Comment (abarth) [inline packages/flutter/lib/src/material/date_picker.dart]: Rather than precomputing this information in build, I'd recommend computing it on-the-fly when we try to vibrate.  A State object has a `context` property that you can feed to `Theme.of` whenever you like.

Comment (xster) [inline packages/flutter/lib/src/material/date_picker.dart]: For my own education, what's the difference? This only changes when set state is called I assume? If it were possible to keep the modal dialog on while changing the platform in the gallery, would it indirectly trigger a set state or no?

Comment (abarth) [inline packages/flutter/lib/src/material/date_picker.dart]: > For my own education, what's the difference?

There's very little operational difference.  The general pattern is that build() functions should be free of side effects, and writing the _vibrate field is a side effect, so it's better code hygiene to do it the other way.

> This only changes when set state is called I assume?

If you're called Theme.of(context) since the last build, then build() will be called if the theme data changes.  (Technically it uses a lower-level mechanism than setState, but the effect is the same.)

> If it were possible to keep the modal dialog on while changing the platform in the gallery, would it indirectly trigger a set state or no?

That is possible.  There's a command in the hot reload command line tool to flip the target platform, which you can do while the dialog is on screen.  (Or the all can just do it on a timer whenever it likes.)

When that happens, that will trigger build() to be called again if that widget has called Theme.of(context) since the last time the framework called build().

Basically, we keep track of all the inherited widgets that you've read since build() and make sure to call build() again if any of them change.  If you don't read from an inherited widget the second time around, we won't call build() again when it changes.

In summary, the code in this patch is correct.  It just violates a code hygiene/style guideline.  We try to make the framework code as clean as possible under the assumption that many people will use the framework code as a template for their own code, and it would be better if they copy/pasted the other pattern.

Comment (cbracken) [inline packages/flutter/lib/src/material/date_picker.dart]: TIL - domo arigato! Replaced in #8337.
